### ISSUE TITLE
Builtin functions should behave differently to Python functions when used as a class method

### ISSUE THREAD

**Issue by dpgeorge** [micropython/micropython#1326]
This issue with uPy was originally discovered by @pfalcon and raised on python-dev, with an answer given: https://mail.python.org/pipermail/python-dev/2015-March/138950.html

As an example, the following works in CPython but not uPy:

``` python
class A:
    f_user = lambda self, x: x
    f_native = abs

print(A().f_user)
print(A().f_user(-1.2))
print(A().f_native)
print(A().f_native(-1.2))
```

The issue is that native functions (when used as a method) should behave "magically" as staticmethods, but uPy treats them the same as Python functions and passes the self object as first argument.

We probably should eventually make this work correctly (since @pfalcon hit it and I've now independently hit it).  The main issue is how to make it work elegantly.

**Comment by dpgeorge** [comment]
Here is a "fix" for this issue which is simple and works, but is a bit of a hack:
```diff
--- a/py/runtime.c
+++ b/py/runtime.c
@@ -959,7 +959,10 @@ void mp_convert_member_lookup(mp_obj_t self, const mp_obj_type_t *type, mp_obj_t
             dest[0] = mp_obj_new_checked_fun(type, member);
         } else
         #endif
-        {
+        if (type->make_new == &mp_obj_instance_make_new && mp_obj_get_type(member) == &mp_type_fun_builtin) {
+            // builtin functions should not bind self
+            dest[0] = member;
+        } else {
             // return a bound method, with self being this object
             dest[0] = member;
             dest[1] = self;
```
Not going to apply this patch right now, it's just for the record.

**Comment by dpgeorge** [comment]
Fixed by 332d83343fb3ef5d2b94b4f058aa53fd0493779e
